Add stateful mealyS

axman6 opened the pull request:
Add a `mealyS :: ... (i -> State s o) -> s -> Signal dom i -> Signal dom o` function to `Clash.Explicit.Mealy` and 
`Clash.Prelude.Mealy`, along with doctests showing how lenses can be used for defining the state modifications.

I've also updated the tutorial to use this, as it simplifies the example a lot and avoids the extra plumbing needed to fix the order of arguments in `macS`.

The dependency on mtl (`Control.Monad.State.Strict`) could easily be a dependency on transformers, (`Control.Monad.Trans.State.Strict`), I just added the one I seem more frequently but happy to change if needed.

## Still TODO:

  - [x] Doctests currently fail for un-logged reasons
  - [x] Update tutorial
  - [x] Consider if it makes sense to have a Unbundled `mealySB`
  - [ ] Consider if it makes sense to have a Moore equivalent
  - [x] Write a changelog entry (see changelog/README.md)
  - [x] Check copyright notices are up to date in edited files

axman6 commented on clash-prelude/src/Clash/Explicit/Mealy.hs:
```diff
@@ -27,17 +28,56 @@ import           Clash.Explicit.Signal
+  let delayTop :: Clock System -> Reset System -> Enable System -> (Signal System Int -> Signal System (Maybe Int))
+      topDelay = P.undefined
+:}
+>>> let hello = "hello"
```
Will be removed, currently trying to debug why the doctests are failing to compile the setup for `topDelay`, which should be defined as commented below.

axman6 commented on clash-prelude/src/Clash/Explicit/Mealy.hs:
```diff
@@ -27,17 +28,56 @@ import           Clash.Explicit.Signal
+      topDelay = P.undefined
+:}
+>>> let hello = "hello"
+>>> :{
+let delayS :: Int -> State DelayState (Maybe Int)
+    delayS n = do
+      _remaining <- use untilValid
+      return Nothing
+:}
+-}
+{-
+      -- if remaining > 0
+      -- then do
+      --   untilValid -= 1
+      --   delayed   %= (n +>>)
+      --   return Nothing
+      -- else do
+      --   out     <- uses delayed C.last
+      --   delayed %= (n +>>)
+      --   return (Just out)
+>>> :{
+  let delayTop :: Clock System -> Reset System -> Enable System -> (Signal System Int -> Signal System (Maybe Int))
+      delayTop clk rst en = mealyS clk rst en delayS initialDelayState
+:}
```
This is what I want the setup to run but it's failing with no useful errors.

axman6 commented on clash-prelude/src/Clash/Explicit/Mealy.hs:
```diff
@@ -106,6 +146,65 @@ mealy clk rst en f iS =
+--
+-- delayS :: Int -> State DelayState (Maybe Int)
+-- delayS n = do
+--   remaining <- get untilValid
```
Should be `use`.

axman6 commented on clash-prelude/src/Clash/Prelude/Mealy.hs:
```diff
@@ -29,16 +30,45 @@ import qualified Clash.Explicit.Mealy as E
+        delayed %= (n +>>)
+        return (Just out)
+
+:}
```
Needs to be updated once the Explicit version compiles.

axman6 reviewed (commented):
Some notes on the WIP

axman6 commented:
I've run into some issues with this PR that I'm struggling to track down. I have an extended version of the problem I'm seeing with both the `mealy` and my `mealyS` functions:

```haskell
{-# LANGUAGE DataKinds, TypeApplications, DeriveGeneric, DeriveAnyClass #-}


import Clash.Explicit.Prelude as C
-- import Clash.Explicit.Mealy (mealyS)
import qualified Data.List as L
import Control.Lens (Lens', (%=), (-=), uses, use)
import Control.Monad.State.Strict (State, runState)

data DelayState = DelayState { _history :: Vec 4 Int , _untilValid :: Index 4 } deriving (Generic,NFDataX)

history :: Lens' DelayState (Vec 4 Int)
history f = \(DelayState d u) -> (`DelayState` u) <$> f d

untilValid :: Lens' DelayState (Index 4)
untilValid f = \(DelayState d u) -> DelayState d <$> f u

type Out = (Maybe Int, Vec 4 Int, Vec 4 Int, Index 4, Index 4)

delayS :: Int -> State DelayState Out
delayS n = do
  remaining <- use untilValid
  hist <- use history -- debugging
  if remaining /= 0
  then do
    history    %= (n +>>)
    untilValid -= 1
    -- debugging
    histAfter  <- use history
    remAfter    <- use untilValid
    return (Nothing, hist, histAfter, remaining, remAfter)
  else do
    out        <- uses history C.last
    history    %= (n +>>)
    -- debugging
    histAfter  <- use history
    remAfter   <- use untilValid
    return (Just out, hist, histAfter, remaining, remAfter)

initialDelayState = DelayState (C.repeat 7) maxBound

-- From PR - `mealy` using State
mealyS
  :: ( KnownDomain dom
     , NFDataX s )
  => Clock dom
  -- ^ 'Clock' to synchronize to
  -> Reset dom
  -> Enable dom
  -- ^ Global enable
  -> (i -> State s o)
  -- ^ Transfer function in mealy machine handling inputs using @Control.Monad.Strict.State s@.
  -> s
  -- ^ Initial state
  -> (Signal dom i -> Signal dom o)
  -- ^ Synchronous sequential function with input and output matching that
  -- of the mealy machine
mealyS clk rst en f iS =
  \i -> let (o,s') = unbundle $ (runState . f) <$> i <*> s
            s      = register clk rst en iS s'
        in o
{-# INLINABLE mealyS #-}


-- >>> P.mapM_ print $ L.take 7 $ simulate (topEntity systemClockGen systemResetGen enableGen) [1,2,3,4,5,6,7,8]
-- (Nothing,7 :> 7 :> 7 :> 7 :> Nil,1 :> 7 :> 7 :> 7 :> Nil,3,2)
-- (Nothing,1 :> 7 :> 7 :> 7 :> Nil,2 :> 1 :> 7 :> 7 :> Nil,3,2)
-- (Nothing,2 :> 1 :> 7 :> 7 :> Nil,3 :> 2 :> 1 :> 7 :> Nil,2,1)
-- (Nothing,3 :> 2 :> 1 :> 7 :> Nil,4 :> 3 :> 2 :> 1 :> Nil,1,0)
-- (Just 1,4 :> 3 :> 2 :> 1 :> Nil,5 :> 4 :> 3 :> 2 :> Nil,0,0)
-- (Just 2,5 :> 4 :> 3 :> 2 :> Nil,6 :> 5 :> 4 :> 3 :> Nil,0,0)
-- (Just 3,6 :> 5 :> 4 :> 3 :> Nil,7 :> 6 :> 5 :> 4 :> Nil,0,0)
--
-- Actual result
-- clashi> P.mapM_ print $ L.take 7 $ simulate (topEntity systemClockGen systemResetGen enableGen) [1,2,3,4,5,6,7,8]
-- (Nothing,7 :> 7 :> 7 :> 7 :> Nil,1 :> 7 :> 7 :> 7 :> Nil,3,2)
-- (Nothing,7 :> 7 :> 7 :> 7 :> Nil,2 :> 7 :> 7 :> 7 :> Nil,3,2) -- note 1 is missing
-- (Nothing,2 :> 7 :> 7 :> 7 :> Nil,3 :> 2 :> 7 :> 7 :> Nil,2,1)
-- (Nothing,3 :> 2 :> 7 :> 7 :> Nil,4 :> 3 :> 2 :> 7 :> Nil,1,0)
-- (Just 7,4 :> 3 :> 2 :> 7 :> Nil,5 :> 4 :> 3 :> 2 :> Nil,0,0)
-- (Just 2,5 :> 4 :> 3 :> 2 :> Nil,6 :> 5 :> 4 :> 3 :> Nil,0,0)
-- (Just 3,6 :> 5 :> 4 :> 3 :> Nil,7 :> 6 :> 5 :> 4 :> Nil,0,0)
topEntity :: Clock System -> Reset System -> Enable System -> Signal System Int -> Signal System Out
topEntity clk rst en = mealyS clk rst en delayS initialDelayState

-- Also happens with mealy:
-- clashi> P.mapM_ print $ L.take 7 $ simulate (topEntity' systemClockGen systemResetGen enableGen) [1,2,3,4,5,6,7,8]
-- (Nothing,7 :> 7 :> 7 :> 7 :> Nil,1 :> 7 :> 7 :> 7 :> Nil,3,2)
-- (Nothing,7 :> 7 :> 7 :> 7 :> Nil,2 :> 7 :> 7 :> 7 :> Nil,3,2) -- 1 also missing in second loop
-- (Nothing,2 :> 7 :> 7 :> 7 :> Nil,3 :> 2 :> 7 :> 7 :> Nil,2,1)
-- (Nothing,3 :> 2 :> 7 :> 7 :> Nil,4 :> 3 :> 2 :> 7 :> Nil,1,0)
-- (Just 7,4 :> 3 :> 2 :> 7 :> Nil,5 :> 4 :> 3 :> 2 :> Nil,0,0)
-- (Just 2,5 :> 4 :> 3 :> 2 :> Nil,6 :> 5 :> 4 :> 3 :> Nil,0,0)
-- (Just 3,6 :> 5 :> 4 :> 3 :> Nil,7 :> 6 :> 5 :> 4 :> Nil,0,0)
topEntity' :: Clock System -> Reset System -> Enable System -> Signal System Int -> Signal System Out
topEntity' clk rst en = mealy clk rst en (\s i -> let (o, s') = runState (delayS i) s in (s',o) ) initialDelayState
```
Somehow the first iteration of the simulation seems to somehow reset part of the state, with everything working as expected after that.

rowanG077 commented:
I didn't look at the code in depth. But this is probably is an artifact of  `resetGen` keeping reset high for the first clock cycle. See the example below with `resetGen` vs `unsafeFromHighPolarity $ pure False`.

```Haskell
ghci> import qualified Data.List as L
ghci> import Clash.Explicit.Prelude
ghci> foo clk rst en s = register clk rst en (undefined :: Bool) s
ghci> showX <$> L.take 10 $ simulate (foo systemClockGen resetGen enableGen) (L.repeat True)
"[undefined,undefined,True,True,True,True,True,True,True,True]"
ghci> showX <$> L.take 10 $ simulate (foo systemClockGen (unsafeFromHighPolarity $ pure False) enableGen) (L.repeat True)
"[undefined,True,True,True,True,True,True,True,True,True]"
```

axman6 commented:
Hmmm, that seems like a likely explanation! Any advice on how to work around that?

rowanG077 commented:
If `mealyS` behaves the same as  `mealy` then I'd think this is exactly what you want?

axman6 commented:
Thanks @rowanG077, with that bit of insight I now have working doctests. I do find the discrepancy between the Prelude and Explicit modules a bit disconcerting, is there a good reason for the difference in behaviour? 

Other than the asymmetry between the two version's tests, I think this is ready to go (assuming the idea is a acceptable).

rowanG077 commented:
This is a bug. 

```
ghci> import qualified Data.List as L
ghci> import qualified Clash.Explicit.Prelude as E
ghci> import Clash.Prelude
ghci> foo clk rst en s = E.register clk rst en (undefined :: Bool) s
ghci> showX <$> L.take 10 $ simulate (foo systemClockGen resetGen enableGen) (L.repeat True)
"[undefined,True,True,True,True,True,True,True,True,True]"
ghci> showX <$> L.take 10 $ E.simulate (foo systemClockGen resetGen enableGen) (L.repeat True)
"[undefined,undefined,True,True,True,True,True,True,True,True]
```

christiaanb commented:
@kloonbot run_ci

alex-mckenna commented:
@kloonboot run_ci 1cbfc58

axman6 commented:
> @kloonbot run_ci [1cbfc58](https://github.com/clash-lang/clash-compiler/commit/1cbfc58e6ad0dbd8a64a2016e6c8ef5ae1d795ca)

Any chance of one more CI run @alex-mckenna? Hoping that's sorted the haddock issue (took me ages to track down where the functions were being being imported and exported from!). Also is there any docs about tests that should be run locally before CI?  It'd be nice to not to need someone else in the loop of checking what I've broken.

alex-mckenna commented:
Sure thing

@kloonbot run_ci 026d78a

axman6 commented:
Thanks @DigitalBrains1, yeah it looks like I've resolved all my issues. 

The PR could probably do with a review, I've made the changes I could find that made sense, but haven't done things like add tests (I didn't see any for the other mealy functions, probably because they're so simple and obviously correct). I'm not sure if there is other documentation needed, or things like _since_ annotations, etc.

martijnbastiaan commented:
@kloonbot run_ci 0d913844e25ac1de29e42d06e3bdfe69b1b3f9f7

DigitalBrains1 commented:
I'm sorry about the spurious review (that I have subsequently deleted, but the notification e-mails were probably already sent). I decided against it, we "can fix it in post" as they say in filming. However, who would have guessed, the "Cancel review" button asks in a popup whether you are sure you want to lose all your changes and then _submits the review anyway_. What a stellar feature.

Again, sorry for the noise. I'll have some actually good comments next week I think and hope, but these were not it.

DigitalBrains1 commented:
I'm sorry about the spurious review (that I have subsequently deleted, but the notification e-mails were probably already sent). I decided against it, we "can fix it in post" as they say in filming. However, who would have guessed, the "Cancel review" button asks in a popup whether you are sure you want to lose all your changes and then _submits the review anyway_. What a stellar feature.

Again, sorry for the noise. I'll have some actually good comments next week I think and hope, but these were not it.

DigitalBrains1 commented on clash-prelude/src/Clash/Explicit/Mealy.hs:
```diff
@@ -106,6 +150,63 @@ mealy clk rst en f iS =
+-- a mealy machine using the state monad. This can be particularly useful
+-- when combined with lenses or optics to replicate imperative algorithms.
+--
+-- @
+-- data DelayState = DelayState
+--   { _history    :: Vec 4 Int
+--   , _untilValid :: Index 4
+--   }
+-- makeLenses 'DelayedState
+--
+-- initialDelayState = DelayState (repeat 0) maxBound
+--
+-- delayS :: Int -> State DelayState (Maybe Int)
+-- delayS n = do
+--   history   %= (n +>>)
+--   remaining <- use untilValid
+--   if remaining > 0
+--   then do
+--      remaining -= 1
+--      return Nothing
+--    else do
+--      out <- uses history last
+--      return (Just out)
+--
+-- delayTop ::'KnownDomain' dom
+--   => 'Clock' dom
+--   -> 'Reset' dom
+--   -> 'Enable' dom
+--   -> ('Signal' dom Int -> 'Signal' dom (Maybe Int))
+-- delayTop clk rst en = 'mealyS' clk rst en delayS initialDelayState
+-- @
```
```suggestion
-- @
-- data DelayState = DelayState
--   { _history    :: Vec 4 Int
--   , _untilValid :: Index 4
--   }
--   deriving (Generic, NFDataX)
-- makeLenses ''DelayState
--
-- initialDelayState = DelayState (repeat 0) maxBound
--
-- delayS :: Int -> State DelayState (Maybe Int)
-- delayS n = do
--   history   %= (n +>>)
--   remaining <- use untilValid
--   if remaining > 0
--   then do
--      untilValid -= 1
--      return Nothing
--    else do
--      out <- uses history last
--      return (Just out)
--
-- delayTop ::'KnownDomain' dom
--   => 'Clock' dom
--   -> 'Reset' dom
--   -> 'Enable' dom
--   -> ('Signal' dom Int -> 'Signal' dom (Maybe Int))
-- delayTop clk rst en = 'mealyS' clk rst en delayS initialDelayState
-- @
```

DigitalBrains1 commented on clash-prelude/src/Clash/Prelude/Mealy.hs:
```diff
@@ -129,6 +172,54 @@ mealyB
+-- a mealy machine using the state monad. This can be particularly useful
+-- when combined with lenses or optics to replicate imperative algorithms.
+--
+-- @
+-- data DelayState = DelayState
+--   { _history    :: Vec 4 Int
+--   , _untilValid :: Index 4
+--   }
+-- makeLenses 'DelayedState
+--
+-- initialDelayState = DelayState (repeat 0) maxBound
+--
+-- delayS :: Int -> State DelayState (Maybe Int)
+-- delayS n = do
+--   history   %= (n +>>)
+--   remaining <- use untilValid
+--   if remaining > 0
+--   then do
+--      remaining -= 1
+--      return Nothing
+--    else do
+--      out <- uses history last
+--      return (Just out)
+--
+-- delayTop :: HiddenClockResetEnable dom  => 'Signal' dom Int -> 'Signal' dom (Maybe Int)
+-- delayTop = 'mealyS' delayS initialDelayState
+-- @
```
```suggestion
-- @
-- data DelayState = DelayState
--   { _history    :: Vec 4 Int
--   , _untilValid :: Index 4
--   }
--   deriving (Generic, NFDataX)
-- makeLenses ''DelayState
--
-- initialDelayState = DelayState (repeat 0) maxBound
--
-- delayS :: Int -> State DelayState (Maybe Int)
-- delayS n = do
--   history   %= (n +>>)
--   remaining <- use untilValid
--   if remaining > 0
--   then do
--      untilValid -= 1
--      return Nothing
--    else do
--      out <- uses history last
--      return (Just out)
--
-- delayTop :: HiddenClockResetEnable dom  => 'Signal' dom Int -> 'Signal' dom (Maybe Int)
-- delayTop = 'mealyS' delayS initialDelayState
-- @
```

DigitalBrains1 replied:
Because GitHub doesn't really show it, I'll point out what this actually boils down to:
```diff
diff --git a/clash-prelude/src/Clash/Explicit/Mealy.hs b/clash-prelude/src/Clash/Explicit/Mealy.hs
index 08b23ca89..022a9e43f 100644
--- a/clash-prelude/src/Clash/Explicit/Mealy.hs
+++ b/clash-prelude/src/Clash/Explicit/Mealy.hs
@@ -159,7 +159,8 @@ mealy clk rst en f iS =
 --   { _history    :: Vec 4 Int
 --   , _untilValid :: Index 4
 --   }
--- makeLenses 'DelayedState
+--   deriving (Generic, NFDataX)
+-- makeLenses ''DelayState
 --
 -- initialDelayState = DelayState (repeat 0) maxBound
 --
@@ -169,7 +170,7 @@ mealy clk rst en f iS =
 --   remaining <- use untilValid
 --   if remaining > 0
 --   then do
---      remaining -= 1
+--      untilValid -= 1
 --      return Nothing
 --    else do
 --      out <- uses history last
```
and then the same for the other file.

DigitalBrains1 reviewed (commented):
I really like this addition! Actually providing some basic pointers about using lenses and the state monad is very helpful to get users going.

I noticed that you worked on your examples a bit more but forgot to copy the updated code to the actual example in Haddock. These suggestions should fix that; I verified they actually work (if you put it in a `.hs` file and add some import statements).

axman6 commented:
Thanks heaps for the review @DigitalBrains1, I've incorporated your changes, I had indeed forgotten to go back and update them (getting the doctests was such a pain I completely forgot about the actual haddocks). 

I've now also added the `mealySB` variants which handle bundling, I've kept their docs brief because it should be fairly obvious from the context of `mealyS` and `mealyB`.

DigitalBrains1 commented on clash-prelude/src/Clash/Explicit/Mealy.hs:
```diff
@@ -149,3 +254,23 @@ mealyB
+  -> Reset dom
+  -> Enable dom
+  -> (i -> State s o)
+  -- ^ Transfer function in mealy machine form: @state -> input -> (newstate,output)@
```
```suggestion
  -- ^ Transfer function in mealy machine handling inputs using @Control.Monad.Strict.State s@.
```

DigitalBrains1 reviewed (changes requested):
Oh yes, the doctests can be a real pain. Usually with extensions, but especially the fact that there is no way to do `NoImplicitPrelude`, whereas that is a basic tenet of Clash... :-(

It is a good idea to add `mealySB`, thanks!

I don't think this is the right moment to introduce a linter pragma... I feel if we start doing that, we need to be more systematic about it, not just drop a single one here. I suggest we don't do it in this PR.

axman6 commented:
Oh sorry about that @DigitalBrains1, that was left in by mistake. If you suggest the change I'll accept it, I'm away from my computer at the moment.

DigitalBrains1 commented on clash-prelude/src/Clash/Explicit/Mealy.hs:
```diff
@@ -15,29 +16,75 @@
 {-# LANGUAGE NoGeneralizedNewtypeDeriving #-}
 
 {-# LANGUAGE Safe #-}
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+{-# HLINT ignore "Redundant lambda" #-}
```
```suggestion
```

DigitalBrains1 commented:
@kloonbot run_ci 51bb400

DigitalBrains1 commented:
I see you still have a TODO note open:
> Consider if it makes sense to have a Moore equivalent

Personally, I think it's not needed because writing a Moore circuit with `State` is easy to accomplish using `register` as the memory element (you could look at the UART in `Clash.Examples` although that module could use some tender love and care. It does show how easy it is.)

I was going to merge this and then noticed the unfinished task. Are you happy with the current state, can I merge it?

DigitalBrains1 commented:
Also, this is going to be a _Squash and merge_ style merge. Do you have a preference regarding commit message or can I cobble something together from your PR cover letter? :-)

axman6 commented:
Yep happy to merge, that task was just there to spark discussion but the more so thought about it the less it made sense. Also happy with squash and merge, with whatever message is appropriate for the repo. Thanks for all your persistence!

DigitalBrains1 commented:
> Thanks for all your persistence!

Hahaha, I should be the one saying that to you :-). Thanks for the contribution!
